exporter/tanzuobservabilityexporter/transformer.go

```diff
@@ -0,0 +1,228 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tanzuobservabilityexporter
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/wavefronthq/wavefront-sdk-go/senders"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+	tracetranslator "go.opentelemetry.io/collector/translator/trace"
+)
+
+type traceTransformer struct {
+	ResourceAttributes pdata.AttributeMap
+	Config             *Config
+}
+
+func newTraceTransformer(resource pdata.Resource, cfg *Config) *traceTransformer {
+	t := &traceTransformer{
+		ResourceAttributes: resource.Attributes(),
+		Config:             cfg,
+	}
+	return t
+}
+
+var (
+	errInvalidSpanID  = errors.New("SpanID is invalid")
+	errInvalidTraceID = errors.New("TraceID is invalid")
+)
+
+type Span struct {
+	Name           string
+	TraceID        uuid.UUID
+	SpanID         uuid.UUID
+	ParentSpanID   uuid.UUID
+	Tags           map[string]string
+	StartMillis    int64
+	DurationMillis int64
+	SpanLogs       []senders.SpanLog
+}
+
+func (t *traceTransformer) Span(orig pdata.Span) (Span, error) {
+	traceID, err := traceIDtoUUID(orig.TraceID())
+	if err != nil {
+		return Span{}, errInvalidTraceID
+	}
+
+	spanID, err := spanIDtoUUID(orig.SpanID())
+	if err != nil {
+		return Span{}, errInvalidSpanID
+	}
+
+	parentSpanID, err := parentSpanIDtoUUID(orig.ParentSpanID())
+	if err != nil {
+		return Span{}, errInvalidSpanID
+	}
+
+	startMillis, durationMillis := calculateTimes(orig)
+
+	tags := attributesToTags(t.ResourceAttributes, orig.Attributes())
+	t.setRequiredTags(tags)
+
+	tags[labelSpanKind] = spanKind(orig)
+
+	errorTags := errorTagsFromStatus(orig.Status())
+	for k, v := range errorTags {
+		tags[k] = v
+	}
+
+	if len(orig.TraceState()) > 0 {
+		tags[tracetranslator.TagW3CTraceState] = string(orig.TraceState())
+	}
+
+	return Span{
+		Name:           orig.Name(),
+		TraceID:        traceID,
+		SpanID:         spanID,
+		ParentSpanID:   parentSpanID,
+		Tags:           tags,
+		StartMillis:    startMillis,
+		DurationMillis: durationMillis,
+		SpanLogs:       eventsToLogs(orig.Events()),
+	}, nil
+}
+
+func spanKind(span pdata.Span) string {
+	switch span.Kind() {
+	case pdata.SpanKindCLIENT:
+		return "client"
+	case pdata.SpanKindSERVER:
+		return "server"
+	case pdata.SpanKindPRODUCER:
+		return "producer"
+	case pdata.SpanKindCONSUMER:
+		return "consumer"
+	case pdata.SpanKindINTERNAL:
+		return "internal"
+	case pdata.SpanKindUNSPECIFIED:
+		return "unspecified"
+	default:
+		return "unknown"
+	}
+}
+
+func (t *traceTransformer) setRequiredTags(tags map[string]string) {
+	if _, ok := tags[labelService]; !ok {
+		if _, svcNameOk := tags[conventions.AttributeServiceName]; svcNameOk {
+			tags[labelService] = tags[conventions.AttributeServiceName]
```

Comment:
You can reuse the value here
```suggestion
		if svcName, svcNameOk := tags[conventions.AttributeServiceName]; svcNameOk {
			tags[labelService] = svcName
```